libr/bin/format/pe/pe.c

```diff
@@ -838,6 +839,69 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t)* bin) {
 	return computed_cs;
 }
 
+static char* PE_(bin_pe_get_claimed_authentihash)(struct PE_(r_bin_pe_obj_t)* bin) {
+	if (!bin->spcinfo) {
+		return NULL;
+	}
+	RASN1Binary *digest = bin->spcinfo->messageDigest.digest;
+	return r_hex_bin2strdup (digest->binary, digest->length);
+}
+
+static char* PE_(bin_pe_get_actual_authentihash)(struct PE_(r_bin_pe_obj_t)* bin) {
+	const char *hashalgo;
+	const ut8 *block, *hash = NULL;
+	char *hashstr = NULL;
+	ut32 checksum_paddr, security_entry_offset, security_dir_size;
+	ut64 len;
+	PE_(image_data_directory) * data_dir_security;
+	PE_DWord security_dir_offset;
+	RBuffer *buf;
+	RHash *ctx;
+	if (!bin->spcinfo) {
+		return NULL;
+	}
+
+	hashalgo = bin->spcinfo->messageDigest.digestAlgorithm.algorithm->string;
+	checksum_paddr = bin->nt_header_offset + 4 + sizeof(PE_(image_file_header)) + 0x40;
+	security_entry_offset =  bin->nt_header_offset + sizeof(PE_(image_nt_headers)) - 96;
+	data_dir_security = &bin->data_directory[PE_IMAGE_DIRECTORY_ENTRY_SECURITY];
+	security_dir_offset = bin_pe_rva_to_paddr (bin, data_dir_security->VirtualAddress);
+	security_dir_size = data_dir_security->Size;
+
+	buf = r_buf_new ();
+	r_buf_append_buf_slice (buf, bin->b, 0, checksum_paddr);
+	r_buf_append_buf_slice (buf, bin->b,
+		checksum_paddr + 4,
+		security_entry_offset - checksum_paddr - 4);
+	r_buf_append_buf_slice (buf, bin->b, 
+		security_entry_offset + 8,
+		security_dir_offset - security_entry_offset - 8);
+	r_buf_append_buf_slice (buf, bin->b,
+		security_dir_offset + security_dir_size, 
+		r_buf_size (bin->b) - security_dir_offset - security_dir_size);
+	block = r_buf_data (buf, &len);
+	if (!strcmp (hashalgo, "md5")) {
```

Comment:
that's a bit hardcoded and it will be good to support multiple hashes here, not just sha1 and md5, i think there's an api in RHash to get the hash bits from the string name

Reply:
Thanks for the info! I will refactor it to use the RHash api, but still only support `md5` and `sha1` since the Authenticode only supports them.